Problem with midi-components-0.0.js

### Bug Description

The following ``git diff`` is required for the Behringer DDM4000 in order to work and might be required for all midi devices on FreeBSD 13.2-RC3.
```
diff --git a/res/controllers/midi-components-0.0.js b/res/controllers/midi-components-0.0.js
index 6bed8a8bf2..8631be97a9 100644
--- a/res/controllers/midi-components-0.0.js
+++ b/res/controllers/midi-components-0.0.js
@@ -123,7 +123,7 @@
                 undefined !== this.outKey &&
                 undefined !== this.output &&
                 typeof this.output === "function") {
-                this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
+                //this.connections[0] = engine.makeConnection(this.group, this.outKey, this.output.bind(this));
             }
         },
         disconnect: function() {
```

### Version

_No response_

### OS

_No response_

Sorry I was just trying to refer to the topic, I understand that coding is required,
When I start mixxx with the original coding I receive the following error:

```
Uncaught exception: file:///usr/local/share/mixxx/controllers/midi-components-0.0.js:126: Error: script tried to connect to ControlObject ([Channel1], ) which is non-existent.
Backtrace: connect@file:///usr/local/share/mixxx/controllers/midi-components-0.0.js:126
Component@file:///usr/local/share/mixxx/controllers/midi-components-0.0.js:50
Button@file:///usr/local/share/mixxx/controllers/midi-components-0.0.js:163
createComponent@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1568
@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1394
createDeck@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1391
@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1366
notify@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:293
schedule@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:287
throttle@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:112
@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1365
@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1364
createLayerManager@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1334
init@file:///usr/local/share/mixxx/controllers/Behringer-Extension-scripts.js:1278
```
then it says:
"The mapping for your controller "port-0" is not working properly. The script code needs to be fixed. The functionality provided by this controller mapping will be disabled until the issue has been resolved."

--- a/src/controllers/scripting/controllerscriptenginebase.cpp
+++ b/src/controllers/scripting/controllerscriptenginebase.cpp
@@ -6,6 +6,7 @@
 #include "errordialoghandler.h"
 #include "mixer/playermanager.h"
 #include "moc_controllerscriptenginebase.cpp"
+#include "util/cmdlineargs.h"
 
 ControllerScriptEngineBase::ControllerScriptEngineBase(
         Controller* controller, const RuntimeLoggingCategory& logger)
@@ -13,6 +14,7 @@ ControllerScriptEngineBase::ControllerScriptEngineBase(
           m_pJSEngine(nullptr),
           m_pController(controller),
           m_logger(logger),
+          m_bPedantic(false),
           m_bTesting(false) {
     // Handle error dialog buttons
     qRegisterMetaType<QMessageBox::StandardButton>("QMessageBox::StandardButton");
@@ -22,6 +24,8 @@ bool ControllerScriptEngineBase::initialize() {
     VERIFY_OR_DEBUG_ASSERT(!m_pJSEngine) {
         return false;
     }
+
+    m_bPedantic = CmdlineArgs::Instance().getControllerPedantic();
 
     // Create the Script Engine
     m_pJSEngine = std::make_shared<QJSEngine>(this);

--- a/src/controllers/scripting/controllerscriptenginebase.cpp
+++ b/src/controllers/scripting/controllerscriptenginebase.cpp
@@ -14,7 +14,7 @@ ControllerScriptEngineBase::ControllerScriptEngineBase(
           m_pJSEngine(nullptr),
           m_pController(controller),
           m_logger(logger),
-          m_bPedantic(false),
+          m_bAbortOnWarning(false),
           m_bTesting(false) {
     // Handle error dialog buttons
     qRegisterMetaType<QMessageBox::StandardButton>("QMessageBox::StandardButton");
@@ -25,7 +25,7 @@ bool ControllerScriptEngineBase::initialize() {
         return false;
     }
 
-    m_bPedantic = CmdlineArgs::Instance().getControllerPedantic();
+    m_bAbortOnWarning = CmdlineArgs::Instance().getControllerAbortOnWarning();
 
     // Create the Script Engine
     m_pJSEngine = std::make_shared<QJSEngine>(this);

--- a/src/controllers/scripting/controllerscriptenginebase.h
+++ b/src/controllers/scripting/controllerscriptenginebase.h
@@ -32,6 +32,10 @@ class ControllerScriptEngineBase : public QObject {
     void showScriptExceptionDialog(const QJSValue& evaluationResult, bool bFatal = false);
     void throwJSError(const QString& message);
 
+    bool isPedantic() const {
+        return m_bPedantic;
+    }
+
     inline void setTesting(bool testing) {
         m_bTesting = testing;
     };
@@ -50,6 +54,8 @@ class ControllerScriptEngineBase : public QObject {
 
     Controller* m_pController;
     const RuntimeLoggingCategory m_logger;
+
+    bool m_bPedantic;
 
     bool m_bTesting;
 

--- a/src/controllers/scripting/controllerscriptenginebase.h
+++ b/src/controllers/scripting/controllerscriptenginebase.h
@@ -32,8 +32,8 @@ class ControllerScriptEngineBase : public QObject {
     void showScriptExceptionDialog(const QJSValue& evaluationResult, bool bFatal = false);
     void throwJSError(const QString& message);
 
-    bool isPedantic() const {
-        return m_bPedantic;
+    bool willAbortOnWarning() const {
+        return m_bAbortOnWarning;
     }
 
     inline void setTesting(bool testing) {
@@ -55,7 +55,7 @@ class ControllerScriptEngineBase : public QObject {
     Controller* m_pController;
     const RuntimeLoggingCategory m_logger;
 
-    bool m_bPedantic;
+    bool m_bAbortOnWarning;
 
     bool m_bTesting;
 

--- a/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
+++ b/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
@@ -110,8 +110,7 @@ ControlObjectScript* ControllerScriptInterfaceLegacy::getControlObjectScript(
 double ControllerScriptInterfaceLegacy::getValue(const QString& group, const QString& name) {
     ControlObjectScript* coScript = getControlObjectScript(group, name);
     if (coScript == nullptr) {
-        qCWarning(m_logger) << "Unknown control" << group << name
-                            << ", returning 0.0";
+        logOrThrowError(QStringLiteral("Unknown control (%1, %2) returning 0.0").arg(group, name));
         return 0.0;
     }
     return coScript->get();
@@ -120,8 +119,9 @@ double ControllerScriptInterfaceLegacy::getValue(const QString& group, const QSt
 void ControllerScriptInterfaceLegacy::setValue(
         const QString& group, const QString& name, double newValue) {
     if (util_isnan(newValue)) {
-        qCWarning(m_logger) << "script setting [" << group << ","
-                            << name << "] to NotANumber, ignoring.";
+        logOrThrowError(QStringLiteral(
+                "Script tried setting (%1, %2) to NotANumber (NaN)")
+                                .arg(group, name));
         return;
     }
 
@@ -141,8 +141,7 @@ void ControllerScriptInterfaceLegacy::setValue(
 double ControllerScriptInterfaceLegacy::getParameter(const QString& group, const QString& name) {
     ControlObjectScript* coScript = getControlObjectScript(group, name);
     if (coScript == nullptr) {
-        qCWarning(m_logger) << "Unknown control" << group << name
-                            << ", returning 0.0";
+        logOrThrowError(QStringLiteral("Unknown control (%1, %2) returning 0.0").arg(group, name));
         return 0.0;
     }
     return coScript->getParameter();
@@ -151,8 +150,9 @@ double ControllerScriptInterfaceLegacy::getParameter(const QString& group, const
 void ControllerScriptInterfaceLegacy::setParameter(
         const QString& group, const QString& name, double newParameter) {
     if (util_isnan(newParameter)) {
-        qCWarning(m_logger) << "script setting [" << group << ","
-                            << name << "] to NotANumber, ignoring.";
+        logOrThrowError(QStringLiteral(
+                "Script tried setting (%1, %2) to NotANumber (NaN)")
+                                .arg(group, name));
         return;
     }
 
@@ -170,16 +170,16 @@ void ControllerScriptInterfaceLegacy::setParameter(
 double ControllerScriptInterfaceLegacy::getParameterForValue(
         const QString& group, const QString& name, double value) {
     if (util_isnan(value)) {
-        qCWarning(m_logger) << "script setting [" << group << ","
-                            << name << "] to NotANumber, ignoring.";
+        logOrThrowError(QStringLiteral(
+                "Script tried setting (%1, %2) to NotANumber (NaN)")
+                                .arg(group, name));
         return 0.0;
     }
 
     ControlObjectScript* coScript = getControlObjectScript(group, name);
 
     if (coScript == nullptr) {
-        qCWarning(m_logger) << "Unknown control" << group << name
-                            << ", returning 0.0";
+        logOrThrowError(QStringLiteral("Unknown control (%1, %2) returning 0.0").arg(group, name));
         return 0.0;
     }
 
@@ -197,8 +197,7 @@ double ControllerScriptInterfaceLegacy::getDefaultValue(const QString& group, co
     ControlObjectScript* coScript = getControlObjectScript(group, name);
 
     if (coScript == nullptr) {
-        qCWarning(m_logger) << "Unknown control" << group << name
-                            << ", returning 0.0";
+        logOrThrowError(QStringLiteral("Unknown control (%1, %2) returning 0.0").arg(group, name));
         return 0.0;
     }
 
@@ -210,8 +209,7 @@ double ControllerScriptInterfaceLegacy::getDefaultParameter(
     ControlObjectScript* coScript = getControlObjectScript(group, name);
 
     if (coScript == nullptr) {
-        qCWarning(m_logger) << "Unknown control" << group << name
-                            << ", returning 0.0";
+        logOrThrowError(QStringLiteral("Unknown control (%1, %2) returning 0.0").arg(group, name));
         return 0.0;
     }
 
@@ -249,10 +247,10 @@ QJSValue ControllerScriptInterfaceLegacy::makeConnectionInternal(
     }
 
     if (!callback.isCallable()) {
-        m_pScriptEngineLegacy->throwJSError("Tried to connect (" + group + ", " + name +
-                ")" +
-                " to an invalid callback. Make sure that your code contains no "
-                "syntax errors.");
+        logOrThrowError(QStringLiteral(
+                "Tried to connect (%1, %2) to an invalid callback. Make sure "
+                "that your code contains no syntax errors.")
+                                .arg(group, name));
         return QJSValue();
     }
 
@@ -293,6 +291,10 @@ void ControllerScriptInterfaceLegacy::triggerScriptConnection(
     ControlObjectScript* coScript =
             getControlObjectScript(connection.key.group, connection.key.item);
     if (coScript == nullptr) {
+        logOrThrowError(QStringLiteral(
+                "Script tried to trigger (%1, %2) which is non-existent.")
+                                .arg(connection.key.group,
+                                        connection.key.item));
         return;
     }
 
@@ -311,6 +313,11 @@ QJSValue ControllerScriptInterfaceLegacy::connectControl(const QString& group,
         const QString& name,
         const QJSValue& passedCallback,
         bool disconnect) {
+    logOrThrowError(QStringLiteral(
+            "Script tried to connect to (%1, %2) using `connectControl` which "
+            "is deprecated. Use `makeConnection` instead!")
+                            .arg(group, name));
+
     // The passedCallback may or may not actually be a function, so when
     // the actual callback function is found, store it in this variable.
     QJSValue actualCallbackFunction;
@@ -427,9 +434,13 @@ QJSValue ControllerScriptInterfaceLegacy::connectControl(const QString& group,
 
 void ControllerScriptInterfaceLegacy::trigger(const QString& group, const QString& name) {
     ControlObjectScript* coScript = getControlObjectScript(group, name);
-    if (coScript != nullptr) {
-        coScript->emitValueChanged();
+    if (coScript == nullptr) {
+        logOrThrowError(QStringLiteral(
+                "Script tried to trigger (%1, %2) which is non-existent.")
+                                .arg(group, name));
+        return;
     }
+    coScript->emitValueChanged();
 }
 
 void ControllerScriptInterfaceLegacy::logOrThrowError(const QString& errorMessage) const {
@@ -441,12 +452,15 @@ void ControllerScriptInterfaceLegacy::logOrThrowError(const QString& errorMessag
 }
 
 void ControllerScriptInterfaceLegacy::log(const QString& message) {
-    qCDebug(m_logger) << "engine.log is deprecated. Use console.log instead.";
+    logOrThrowError(QStringLiteral("`engine.log` is deprecated. Use `console.log` instead!"));
     qCDebug(m_logger) << message;
 }
 int ControllerScriptInterfaceLegacy::beginTimer(
         int intervalMillis, QJSValue timerCallback, bool oneShot) {
     if (timerCallback.isString()) {
+        logOrThrowError(
+                QStringLiteral("passed a string to `engine.beginTimer`, please "
+                               "pass a function instead!"));
         timerCallback = m_pScriptEngineLegacy->jsEngine()->evaluate(timerCallback.toString());
     } else if (!timerCallback.isCallable()) {
         QString sErrorMessage(
@@ -475,7 +489,7 @@ int ControllerScriptInterfaceLegacy::beginTimer(
     info.oneShot = oneShot;
     m_timers[timerId] = info;
     if (timerId == 0) {
-        qCWarning(m_logger) << "Script timer could not be created";
+        logOrThrowError(QStringLiteral("Script timer could not be created"));
     } else if (oneShot) {
         qCDebug(m_logger) << "Starting one-shot timer:" << timerId;
     } else {
@@ -486,8 +500,9 @@ int ControllerScriptInterfaceLegacy::beginTimer(
 
 void ControllerScriptInterfaceLegacy::stopTimer(int timerId) {
     if (!m_timers.contains(timerId)) {
-        qCWarning(m_logger) << "Killing timer" << timerId
-                            << ": That timer does not exist!";
+        logOrThrowError(QStringLiteral(
+                "Tried to kill Timer \"%1\" that does not exists")
+                                .arg(timerId));
         return;
     }
     qCDebug(m_logger) << "Killing timer:" << timerId;

--- a/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
+++ b/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
@@ -240,10 +240,10 @@ QJSValue ControllerScriptInterfaceLegacy::makeConnectionInternal(
         // The test setups do not run all of Mixxx, so ControlObjects not
         // existing during tests is okay.
         if (!m_pScriptEngineLegacy->isTesting()) {
-            m_pScriptEngineLegacy->throwJSError(
-                    "script tried to connect to "
-                    "ControlObject (" +
-                    group + ", " + name + ") which is non-existent.");
+            logOrThrowError(
+                    QStringLiteral("script tried to connect to ControlObject "
+                                   "(%1, %2) which is non-existent.")
+                            .arg(group, name));
         }
         return QJSValue();
     }
@@ -429,6 +429,14 @@ void ControllerScriptInterfaceLegacy::trigger(const QString& group, const QStrin
     ControlObjectScript* coScript = getControlObjectScript(group, name);
     if (coScript != nullptr) {
         coScript->emitValueChanged();
+    }
+}
+
+void ControllerScriptInterfaceLegacy::logOrThrowError(const QString& errorMessage) const {
+    if (m_pScriptEngineLegacy->isPedantic()) {
+        m_pScriptEngineLegacy->throwJSError(errorMessage);
+    } else {
+        qCWarning(m_logger) << errorMessage;
     }
 }
 

--- a/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
+++ b/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp
@@ -444,7 +444,7 @@ void ControllerScriptInterfaceLegacy::trigger(const QString& group, const QStrin
 }
 
 void ControllerScriptInterfaceLegacy::logOrThrowError(const QString& errorMessage) const {
-    if (m_pScriptEngineLegacy->isPedantic()) {
+    if (m_pScriptEngineLegacy->willAbortOnWarning()) {
         m_pScriptEngineLegacy->throwJSError(errorMessage);
     } else {
         qCWarning(m_logger) << errorMessage;

--- a/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
+++ b/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h
@@ -82,6 +82,7 @@ class ControllerScriptInterfaceLegacy : public QObject {
             bool skipSuperseded = false);
     QHash<ConfigKey, ControlObjectScript*> m_controlCache;
     ControlObjectScript* getControlObjectScript(const QString& group, const QString& name);
+    void logOrThrowError(const QString& errorMessage) const;
 
     SoftTakeoverCtrl m_st;
 

--- a/src/util/cmdlineargs.cpp
+++ b/src/util/cmdlineargs.cpp
@@ -20,6 +20,7 @@
 CmdlineArgs::CmdlineArgs()
         : m_startInFullscreen(false), // Initialize vars
           m_controllerDebug(false),
+          m_controllerPedantic(false),
           m_developer(false),
           m_safeMode(false),
           m_useVuMeterGL(true),
@@ -193,6 +194,17 @@ bool CmdlineArgs::parse(const QStringList& arguments, CmdlineArgs::ParseMode mod
     parser.addOption(controllerDebug);
     parser.addOption(controllerDebugDeprecated);
 
+    const QCommandLineOption controllerPedantic(
+            QStringLiteral("controller-pedantic"),
+            forUserFeedback ? QCoreApplication::translate("CmdlineArgs",
+                                      "The controller mapping will issue more "
+                                      "aggressive warnings and errors when "
+                                      "detecting misuse of controller APIs. "
+                                      "New Controller Mappings should be "
+                                      "developed with this option enabled!")
+                            : QString());
+    parser.addOption(controllerPedantic);
+
     const QCommandLineOption developer(QStringLiteral("developer"),
             forUserFeedback ? QCoreApplication::translate("CmdlineArgs",
                                       "Enables developer-mode. Includes extra log info, stats on "
@@ -332,6 +344,7 @@ bool CmdlineArgs::parse(const QStringList& arguments, CmdlineArgs::ParseMode mod
 
     m_useVuMeterGL = !(parser.isSet(disableVuMeterGL) || parser.isSet(disableVuMeterGLDeprecated));
     m_controllerDebug = parser.isSet(controllerDebug) || parser.isSet(controllerDebugDeprecated);
+    m_controllerPedantic = parser.isSet(controllerPedantic);
     m_developer = parser.isSet(developer);
     m_safeMode = parser.isSet(safeMode) || parser.isSet(safeModeDeprecated);
     m_debugAssertBreak = parser.isSet(debugAssertBreak) || parser.isSet(debugAssertBreakDeprecated);

--- a/src/util/cmdlineargs.cpp
+++ b/src/util/cmdlineargs.cpp
@@ -20,7 +20,7 @@
 CmdlineArgs::CmdlineArgs()
         : m_startInFullscreen(false), // Initialize vars
           m_controllerDebug(false),
-          m_controllerPedantic(false),
+          m_controllerAbortOnWarning(false),
           m_developer(false),
           m_safeMode(false),
           m_useVuMeterGL(true),
@@ -194,8 +194,8 @@ bool CmdlineArgs::parse(const QStringList& arguments, CmdlineArgs::ParseMode mod
     parser.addOption(controllerDebug);
     parser.addOption(controllerDebugDeprecated);
 
-    const QCommandLineOption controllerPedantic(
-            QStringLiteral("controller-pedantic"),
+    const QCommandLineOption controllerAbortOnWarning(
+            QStringLiteral("controller-abort-on-warning"),
             forUserFeedback ? QCoreApplication::translate("CmdlineArgs",
                                       "The controller mapping will issue more "
                                       "aggressive warnings and errors when "
@@ -203,7 +203,7 @@ bool CmdlineArgs::parse(const QStringList& arguments, CmdlineArgs::ParseMode mod
                                       "New Controller Mappings should be "
                                       "developed with this option enabled!")
                             : QString());
-    parser.addOption(controllerPedantic);
+    parser.addOption(controllerAbortOnWarning);
 
     const QCommandLineOption developer(QStringLiteral("developer"),
             forUserFeedback ? QCoreApplication::translate("CmdlineArgs",
@@ -344,7 +344,7 @@ bool CmdlineArgs::parse(const QStringList& arguments, CmdlineArgs::ParseMode mod
 
     m_useVuMeterGL = !(parser.isSet(disableVuMeterGL) || parser.isSet(disableVuMeterGLDeprecated));
     m_controllerDebug = parser.isSet(controllerDebug) || parser.isSet(controllerDebugDeprecated);
-    m_controllerPedantic = parser.isSet(controllerPedantic);
+    m_controllerAbortOnWarning = parser.isSet(controllerAbortOnWarning);
     m_developer = parser.isSet(developer);
     m_safeMode = parser.isSet(safeMode) || parser.isSet(safeModeDeprecated);
     m_debugAssertBreak = parser.isSet(debugAssertBreak) || parser.isSet(debugAssertBreakDeprecated);

--- a/src/util/cmdlineargs.h
+++ b/src/util/cmdlineargs.h
@@ -34,6 +34,9 @@ class CmdlineArgs final {
     bool getControllerDebug() const {
         return m_controllerDebug;
     }
+    bool getControllerPedantic() const {
+        return m_controllerPedantic;
+    }
     bool getDeveloper() const { return m_developer; }
     bool getSafeMode() const { return m_safeMode; }
     bool useColors() const {
@@ -73,6 +76,7 @@ class CmdlineArgs final {
     QList<QString> m_musicFiles;    // List of files to load into players at startup
     bool m_startInFullscreen;       // Start in fullscreen mode
     bool m_controllerDebug;
+    bool m_controllerPedantic; // Controller Engine will be stricter
     bool m_developer; // Developer Mode
     bool m_safeMode;
     bool m_useVuMeterGL;

--- a/src/util/cmdlineargs.h
+++ b/src/util/cmdlineargs.h
@@ -34,8 +34,8 @@ class CmdlineArgs final {
     bool getControllerDebug() const {
         return m_controllerDebug;
     }
-    bool getControllerPedantic() const {
-        return m_controllerPedantic;
+    bool getControllerAbortOnWarning() const {
+        return m_controllerAbortOnWarning;
     }
     bool getDeveloper() const { return m_developer; }
     bool getSafeMode() const { return m_safeMode; }
@@ -76,7 +76,7 @@ class CmdlineArgs final {
     QList<QString> m_musicFiles;    // List of files to load into players at startup
     bool m_startInFullscreen;       // Start in fullscreen mode
     bool m_controllerDebug;
-    bool m_controllerPedantic; // Controller Engine will be stricter
+    bool m_controllerAbortOnWarning; // Controller Engine will be stricter
     bool m_developer; // Developer Mode
     bool m_safeMode;
     bool m_useVuMeterGL;

